docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,217 @@
+
+## No Goals
+
+* Removing initial startup for Thanos Store Gateway completely as designed in [Cortex, no initial block sync](https://github.com/thanos-io/thanos/issues/1813)
+  * However this proposal might be a step towards that as we might be able to load and build index-cache/index quickly on demand from disk. See [Future Work](./201912_thanos_binary_index_header.md#Future-work)
+  * At the same time be able to load `index-header` at query time directly from bucket is not a goal of this proposal.
+* Decreasing size. While it would be nice to use less space; our aim is latency of building/loading the block. That might be correlated with size, but not necessarily (e.g when extra considering compression)
+
+## Verification
+
+* Compare memory and latency usage for startup without index-header being generated on compactor
+* Benchmark for loading the `index-header` vs `index-cache.json` into Store GW.
+
+## Proposal
+
+TSDB index is in binary [format](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md). 
+
+To allow reduced resource consumption and effort when building (1), (2), "index-header" for blocks we plan to reuse similar format for sections like symbols, label indices and 
+posting offsets in separate the file called `index-header` that will replace currently existing `index-cache.json`.
+ 
+The process for building this will be as follows:
+
+* Thanks to https://github.com/thanos-io/thanos/pull/1792 we can check final size of index and scan for TOC file.
+* With TOC:
+  * Get the symbols table and copy it to the local file.
+  * Get the posting table and copy it to the local file.
+* Write new TOC file at the end of file.
+  
+With that effort building time and resources should be compared with downloading the prebuilt `index-header` from the bucket.
+This allows us to reduce complexity of the system as we don't need to cache that in object storage by compactor.
+
+Thanks to this format we can reuse most of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664) code
+to load file.
+
+Thanos will build/compose all index-headers on the startup for now, however in theory we can load and build those blocks on demand.
+Given the minimal memory that each loaded block should take now, this is described as [Future Work](./201912_thanos_binary_index_header.md#Future-Work)
+
+### Length of Posting to fetch
+
+While idea of combing different pieces of TSDB index as our index-header is great, unfortunately we heavily rely 
+on information about size of each posting represented as `postingRange.End`. 
+
+We need to know that apriori to know how to partition and how many bytes we need to fetch from the storage to get each posting: https://github.com/thanos-io/thanos/blob/7e11afe64af0c096743a3de8a594616abf52be45/pkg/store/bucket.go#L1567
```

Comment:
> We need to know that apriori to know how to partition

I guess you mean `We need to know apriori how to partition`

Reply:
Will fix in next PR (: